src/huggingface_hub/file_download.py

```diff
@@ -1020,7 +1024,19 @@ def hf_hub_download(
                 proxies=proxies,
                 timeout=etag_timeout,
             )
-            _raise_for_status(r)
+            try:
+                _raise_for_status(r)
+            except EntryNotFoundError:
+                print("Error intercepted")
+                commit_hash = r.headers.get(HUGGINGFACE_HEADER_X_REPO_COMMIT)
+                if commit_hash is not None and not legacy_cache_layout:
+                    no_exist_path = os.path.join(
+                        storage_folder, ".no_exist", commit_hash
+                    )
+                    os.makedirs(no_exist_path, exist_ok=True)
+                    print(f"Making folder {no_exist_path}")
+                    (Path(no_exist_path) / relative_filename).touch()
+                raise
```

Comment:
For the problem you mentioned in https://github.com/huggingface/huggingface_hub/pull/986#issuecomment-1212352703, since we have it, it could still be good to cache the ref.

As you said, always good not to duplicate code so In practice I would create a helper function:
```py
def _cache_commit_hash_for_specific_revision(storage_folder:str, revision:str, commit_hash:str) -> None:
    """Cache reference between a revision (tag or branch) and the corresponding commit hash.
    
    Does nothing if `revision` is already a `commit_hash` or reference is already cached.
    """
    if revision != commit_hash:
        ref_path = os.path.join(storage_folder, "refs", revision)
        if not os.path.exists(ref_path):
            os.makedirs(os.path.dirname(ref_path), exist_ok=True)
            with open(ref_path, "w") as f:
                f.write(commit_hash)
```
and call it just before the `raise` and [later in the code](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/file_download.py#L1119)